File: leet_lint_code/lint_code/binary_tree_maximum_path_sum.cpp
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of lint code problem 94.

#ifndef LEETLINTCODE_LINTCODE_BINARYTREEMAXIMUMPATHSUM_CPP
#define LEETLINTCODE_LINTCODE_BINARYTREEMAXIMUMPATHSUM_CPP

#include <string>
#include <vector>
#include <iostream>
#include <map>
#include <math.h>


class TreeNode {
 public:
   TreeNode(int val) {
     this->val = val;
     this->left = this->right = NULL;
   }

   int val;
   TreeNode *left, *right;
};

class BinaryTreeMaximumPathSum {
 public:
   int maxPathSum(TreeNode * root) {
     std::map<TreeNode*, int> mem;
     return maxPathSumHelp(root, mem);
   }

   int maxPathSumHelp(TreeNode* root, std::map<TreeNode*, int> &mem) {
     if (NULL == root) return 0;
     int left_val = std::max(0, maxPathSumHelp2(root->left, mem));
     int right_val = std::max(0, maxPathSumHelp2(root->right, mem));
     int left_child = maxPathSumHelp(root->left, mem);
     int right_child = maxPathSumHelp(root->right, mem);
     int max = left_val + right_val + root->val;
     if (root->left != NULL) max = std::max(max, left_child);
     if (root->right != NULL) max = std::max(max, right_child);
     return max;
   }

   int maxPathSumHelp2(TreeNode* root, std::map<TreeNode*, int> &mem) {
     if (NULL == root) return 0;
     if (mem.find(root) != mem.end()) return mem[root];
     int max = 0;
     max = std::max(max, maxPathSumHelp2(root->left, mem));
     mem[root] = root->val + std::max(max, maxPathSumHelp2(root->right, mem));
     return mem[root];
   }
};

#endif // LEETLINTCODE_LINTCODE_BINARYTREEMAXIMUMPATHSUM_CPP

File: leet_lint_code/leet_code/java/BinaryTreeMaximumPathSum.java
// Copyright © 2018 Hailin Gu. All rights reserved.
// License(GPL)
// Author: Hailin Gu
// This is a answer of leet code problem 124.
// Date: 2019.8.19


import java.util.HashMap;
import java.util.Map;

public class BinaryTreeMaximumPathSum {
    public int maxPathSum(TreeNode root) {
        if (root == null) {
            return 0;
        }

        Map<TreeNode, Integer> memo = new HashMap<>();
        Map<TreeNode, Integer> memoSingle = new HashMap<>();
        int ans = Integer.MIN_VALUE;
        maxPathSumAll(root, memo, memoSingle);
        for (Map.Entry<TreeNode, Integer> entry : memo.entrySet()) {
            ans = Math.max(ans, entry.getValue());
        }

        for (Map.Entry<TreeNode, Integer> entry : memoSingle.entrySet()) {
            ans = Math.max(ans, entry.getValue());
        }

        return ans;
    }

    int maxPathSumAll(TreeNode root, Map<TreeNode, Integer> memo, Map<TreeNode, Integer> memoSingle) {
        if (root == null) {
            return 0;
        }

        if (memo.containsKey(root)) {
            return memo.get(root);
        }

        int sum = root.val;
        int left = maxPathSumSingle(root.left, memoSingle);
        int right = maxPathSumSingle(root.right, memoSingle);
        sum = sum + Math.max(0, left) + Math.max(0, right);
        memo.put(root, sum);

        maxPathSumAll(root.left, memo, memoSingle);
        maxPathSumAll(root.right, memo, memoSingle);

        return sum;
    }

    int maxPathSumSingle(TreeNode root, Map<TreeNode, Integer> memo) {
        if (root == null) {
            return 0;
        }

        if (memo.containsKey(root)) {
            return memo.get(root);
        }

        int sum = root.val;
        int left = maxPathSumSingle(root.left, memo);
        int right = maxPathSumSingle(root.right, memo);

        sum = Math.max(sum + Math.max(0, left), sum + Math.max(0, right));
        memo.put(root, sum);
        return sum;
    }
}
